### PR TITLE
Chapter 10: Updating users

### PR DESCRIPTION
**Pull request by phucynwa**
Chapter 10: Updating users

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,6 +1,15 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :admin_user, only: :destroy
+
+  def index
+    @users = User.all
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: No need. Please remove this duplicate variable

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
+    @users = User.paginate page: params[:page], per_page: 10
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Move the constant to setting.yml

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -13,16 +22,58 @@ def new
...
+        flash[:success] = t ".profile_updated"
+        redirect_to @user
+      else
+        render :edit
+      end
+    end
+  end
+
+  def destroy
+    User.find_by(id: params[:id]).destroy
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: If user is nil/don't find the user with the params[:id], what will happen? Will destroy method be called?

File: app/models/user.rb [hunk 4]
```diff
@@ -33,4 +33,22 @@ def authenticated? remember_token
   def forget
     update_attributes remember_digest: nil
   end
+
+  def User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove duplicate class method

File: app/models/user.rb [hunk 5]
```diff
@@ -33,4 +33,22 @@ def authenticated? remember_token
...
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove duplicate

File: app/models/user.rb [hunk 6]
```diff
@@ -33,4 +33,22 @@ def authenticated? remember_token
...
+    self.remember_token = User.new_token
+    update_attributes remember_digest: User.digest(remember_token)
+  end
+
+  def authenticated? remember_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove this method. It is duplicated.

File: app/models/user.rb [hunk 7]
```diff
@@ -33,4 +33,22 @@ def authenticated? remember_token
...
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove this. It is duplicated.

File: app/views/users/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<%= form_for(@user) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/users/_form.html.erb]: remove ()

File: app/views/users/_user.html.erb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<li>
+  <%= gravatar_for user, size: 50 %>
```
Comment (mdmahmudurrahman) [inline app/views/users/_user.html.erb]: Move constant to settings.yml

File: db/seeds.rb [hunk 10]
```diff
@@ -1,7 +1,8 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+User.create! name: "Example User", email: "example@railstutorial.org", password: "foobar", password_confirmation: "foobar", admin: true
```
Comment (mdmahmudurrahman) [inline db/seeds.rb]: Maximum 80 characters per line.

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -13,16 +22,58 @@ def new
...
+      redirect_to login_url
+    end
+  end
+
+  def correct_user
+    @user = User.find_by id: params[:id]
+    redirect_to root_url unless current_user? @user
+  end
+
+  def admin_user
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Bad naming! Name a function in more precise way depending on it's functionality. 
For example, you can name it verify_admin, verify_admin_user etc

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -13,16 +21,61 @@ def new
...
+      log_in @user
+      flash[:success] = t ".welcome"
       redirect_to @user
     else
       render :new
     end
   end
 
+  def edit
+    @user = User.find_by id: params[:id]
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: If @user is nil, what happen?

Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: If @user is nil, what happen?

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -13,16 +21,61 @@ def new
...
+      end
+    end
+  end
+
+  def destroy
+    user = User.find_by(id: params[:id])
+    unless user.nil?
+      user.destroy
+    end
+    flash[:success] = t ".user_deleted"
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: If user is not deleted, this message will be shown. Isn't it!

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -13,16 +21,64 @@ def new
...
+    end
+  end
+
+  def destroy
+    user = User.find_by(id: params[:id])
+    unless user.nil?
+      user.destroy
+      flash[:success] = t ".user_deleted"
+    end
+    redirect_to users_url
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: So the request will be redirected to users_url for both of the case(delete/not deleted), it that right?

Comment (phucynwa) [reply app/controllers/users_controller.rb]: @mdmahmudurrahman So let it doesn't redirect ?

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -13,16 +21,66 @@ def new
...
+
+  def destroy
+    user = User.find_by(id: params[:id])
+    unless user.nil?
+      user.destroy
+      flash[:success] = t ".user_deleted"
+    else
+      flash[:warning] = t ".user_not_exist"
+    end
+    redirect_to users_url
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]:
```
@user = User.find_by id: params[:id]

if @user
  # do something
else 
 # do another thing
end 
```
Now you can see that, the same logic is duplicated for 4 actions show, edit, update and destroy. Can you use before_action callback to remove this duplicate code.

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -13,16 +19,63 @@ def new
...
+
+  def check_exist_user
+    @user = User.find_by id: params[:id]
+    if @user
+      @user
+    else
+      flash[:warning] = t ".user_not_exist"
+      redirect_to root_url
+    end
+  end
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: You can make it more smaller
```
@user = User.find_by id: params[:id]
return if @user.present? 
flash[:warning] = t ".user_not_exist"
redirect_to root_url
```

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,9 +1,15 @@
...
+  def index
+    @users = User.paginate page: params[:page], per_page: Settings.per_page
+  end
+
   def show
-    @user = User.find_by_id params[:id]
-    if @user.nil?
-      redirect_to root_path
-    end
+    @user
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: No need. before_action callback make it available to the view

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -13,16 +19,63 @@ def new
...
+      log_in @user
+      flash[:success] = t ".welcome"
       redirect_to @user
     else
       render :new
     end
   end
 
+  def edit
+    @user
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Remove it. before_action callback make it available to the view

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (phucynwa) [reply app/controllers/users_controller.rb]: @mdmahmudurrahman Thank you so much.

**Pull request merged**